fluent/src/resource.js

```diff
@@ -1,18 +1,467 @@
-import parse from "./parser";
+import FluentError from "./error.js";
+
+// This regex is used to iterate through the beginnings of messages and terms.
+// With the /m flag, the ^ matches at the beginning of every line.
+const RE_MESSAGE_START = /^(-?[a-zA-Z][a-zA-Z0-9_-]*) *= */mg;
+
+// Both Attributes and Variants are parsed in while loops. These regexes are
+// used to break out of them.
+const RE_ATTRIBUTE_START = /\.([a-zA-Z][a-zA-Z0-9_-]*) *= */y;
+// [^] matches all characters, including newlines.
+// XXX Use /s (dotall) when it's widely supported.
+const RE_VARIANT_START = /\*?\[[^]*?] */y;
+
+const RE_IDENTIFIER = /(-?[a-zA-Z][a-zA-Z0-9_-]*)/y;
+const RE_NUMBER_LITERAL = /(-?[0-9]+(\.[0-9]+)?)/y;
+
+// A "run" is a sequence of text or string literal characters which don't
+// require any special handling. For TextElements such special characters are:
+// { (starts a placeable), \ (starts an escape sequence), and line breaks which
+// require additional logic to check if the next line is indented. For
+// StringLiterals they are: \ (starts an escape sequence), " (ends the
+// literal), and line breaks which are not allowed in StringLiterals. Also note
+// that string runs may be empty, but text runs may not.
+const RE_TEXT_RUN = /([^\\{\n\r]+)/y;
+const RE_STRING_RUN = /([^\\"\n\r]*)/y;
+
+// Escape sequences.
+const RE_UNICODE_ESCAPE = /\\u([a-fA-F0-9]{4})/y;
+const RE_STRING_ESCAPE = /\\([\\"])/y;
+const RE_TEXT_ESCAPE = /\\([\\{])/y;
+
+// Used for trimming TextElements and indents. With the /m flag, the $ matches
+// the end of every line.
+const RE_TRAILING_SPACES = / +$/mg;
+// CRLFs are normalized to LF.
+const RE_CRLF = /\r\n/g;
+
+// Common tokens.
+const TOKEN_BRACE_OPEN = /{\s*/y;
+const TOKEN_BRACE_CLOSE = /\s*}/y;
+const TOKEN_BRACKET_OPEN = /\[\s*/y;
+const TOKEN_BRACKET_CLOSE = /\s*]/y;
+const TOKEN_PAREN_OPEN = /\(\s*/y;
+const TOKEN_ARROW = /\s*->\s*/y;
+const TOKEN_COLON = /\s*:\s*/y;
+// As a deviation from the well-formed Fluent grammar, accept argument lists
+// without commas between arguments.
```

Comment:
I don't think I understand this comment. How's it different from other normalizations?

Reply:
Can you help me re-word the comment? What do you mean by "other normalizations"? The regex under this comment has an optional comma (`.?`) which essentially means that commas in argument lists are treated as whitespace.

Reply:
How does the following sound?

```js
// Note the optional comma. As a deviation from the Fluent EBNF, the parser
// accepts lists of call arguments without commas between them.
const TOKEN_COMMA = /\s*,?\s*/y;
```

Reply:
Ah, I see. I somehow confused the message as "accept argument lists with commas between arguments". Maybe - `As a deviation from the well-formed Fluent grammar, this parser does not enforce commas between arguments.`